Update doc for kafe2go

cverstege opened the pull request:
@JohannesGaessler can you please proofread the changes and maybe help me with the advanced data errors for kafe2go?

JohannesGaessler commented on doc/src/parts/mathematical_foundations.rst:
```diff
@@ -227,33 +240,72 @@ As it turns the method of least squares is a special case of the method
 
-Negative log-likelihood estimator
----------------------------------
+    C_\mathrm{total} = C + \mathrm{constraints}
```
\mathrm{constraints} seems like a weird choice for a variable name in LaTeX.
Is this intentional?

cverstege replied:
I didn't know how to call it, so I just wrote constaints. Do you have any ideas?

JohannesGaessler replied:
I would just use indices for C.
So something like `C_total = C_0 + C_constraint` with some previous hint that `C \equiv C_0`

JohannesGaessler commented on doc/src/parts/mathematical_foundations.rst:
```diff
@@ -227,33 +240,72 @@ As it turns the method of least squares is a special case of the method
+    C_\mathrm{total} = C + \mathrm{constraints}
 
+In general the penalty is given by the current parameter values :math:`\textbf{p}`, the expected
+parameter values :math:`\bm{\mu}` and the covariance matrix :math:`V`:
```
This makes it sound like it's the same covariance matrix as used for the regular cost function.

JohannesGaessler reviewed (changes requested):
Partial review: devloper_guide.rst and mathematical_foundations.rst

cverstege commented on doc/src/parts/mathematical_foundations.rst:
```diff
@@ -50,11 +47,25 @@ correlated errors. In order to make meaningful quantitative
 both components of all uncertainties, expressed in terms of
 their covariance matrix, or by the uncertainties and
-their correlation matrix. 
+their correlation matrix.
```
It says that systematical and statistical uncertainties can be described through independent and correlated erros. Do you have an idea on how to change this/ rewrite that section?

cverstege commented on doc/src/parts/mathematical_foundations.rst:
```diff
@@ -227,33 +240,72 @@ As it turns the method of least squares is a special case of the method
 
+Depending on the chosen cost function, uncertainties are handled differently.
+Simple pointwise gaussian uncertainties can be handled by a negative log likelihood cost function
+with a gaussian distribution for each data point.
+Alternatively the method of least squares can be used which can handle correlations between the
+uncertainties via the covariance or correlation matrix as well.
+The squared pointwise uncertainties correspond to the diagonal elements of the covariance matrix.
 
-Handling uncertainties
-======================
+Uncertainties relative to the data are constant, as the data is constant when performing a fit.
```
How uncertainties work and are handled is imo part of the theoretical foundations. There currently is no other place where we can put this. The User Guide will only contain examples on how to use it, but won't explain how it works in the back end.

cverstege replied:
I thought you meant the whole section. Only the part which cost functions can handle which uncertainties can indeed be moved.

cverstege commented:
@JohannesGaessler I just pushed a new, cleaned up and finished version. It should include all your earlier comments.
After a review of each changed file and fixing I would merge those commits.

JohannesGaessler commented on doc/src/parts/user_guide_kafe2go.rst:
```diff
@@ -0,0 +1,328 @@
+      c: [0.0, 1.0]
+      d: [-5, 5]
+
+Note, that a limited parameter needs a lower and an upper limit. The first value of the list is the
```
```suggestion
Note that a limited parameter needs a lower and an upper limit. The first value of the list is the
```

JohannesGaessler commented on doc/src/parts/user_guide_kafe2go.rst:
```diff
@@ -0,0 +1,328 @@
+
+Note, that a limited parameter needs a lower and an upper limit. The first value of the list is the
+lower limit and the second value is the upper limit.
+If a parameter shall be limited to only positive values, choose 0 as lower and a high value as
```
```suggestion
If a parameter shall be limited to non-negative values, choose 0 as lower and a high value as
```

JohannesGaessler reviewed (changes requested):
I think you misunderstood what I wanted you to write regarding statistical/systematic errors.
I will push a commit with text that I think would be appropriate.

cverstege commented on doc/src/parts/developer_guide.rst:
```diff
@@ -90,3 +90,32 @@ To build the documentation run::
+Modern IDEs can also perform type checking with the help of docstrings.
+That's why, in general, we try to include the types inside the docstrings.
+
+If you're working on the code and find missing, incomplete or hard do understand docstrings,
+please update them and create a `Pull Request <https://github.com/dsavoiu/kafe2/compare>`_ or
+`Issue <https://github.com/dsavoiu/kafe2/issues/new/choose>`_ on GitHub.
```
Removed them.

cverstege commented on doc/src/parts/user_guide_kafe2go.rst:
```diff
@@ -0,0 +1,328 @@
+
+.. code-block:: bash
+
+    kafe2go path/to/fit.yml
```
Maybe we should add a `kafe2go.bat` file if that helps?

JohannesGaessler replied:
That is not the problem.
The problem is that Windows doesn't support shebangs.
On Windows the file extension determines how a file is interpreted.

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -22,11 +22,296 @@ correlations.
+========
+When performing fits with *kafe2*, the data is stored in so-called data containers
+(:py:obj:`~.DataContainerBase`-derived objects).
+The difference between the container classes comes down the type of data they store.
```
```suggestion
The difference between the container classes comes down to the type of data they store.
```

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -22,11 +22,296 @@ correlations.
+
+Setting the data
+----------------
+Data creators are created as regular Python objects from iterables (lists, arrays, etc.) of floats.
```
```suggestion
Data containers are created as regular Python objects from iterables (lists, arrays, etc.) of floats.
```

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -22,11 +22,296 @@ correlations.
+Setting a model function
+------------------------
+
+*kafe2* fit objects accept normal python functions as model functions.
```
```suggestion
*kafe2* fit objects accept normal Python functions as model functions.
```

JohannesGaessler commented on doc/src/parts/mathematical_foundations.rst:
```diff
@@ -227,33 +221,65 @@ As it turns the method of least squares is a special case of the method
+When performing a fit, some values of the model function might have already been determined in
+previous experiments.
+Those results and uncertainties can then be used to constrain the given parameters in a new fit.
+This eradicates the need to manually propagate the uncertainties on the final fit results, as
```
```suggestion
This eliminates the need to manually propagate the uncertainties on the final fit results, as
```

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -22,11 +22,296 @@ correlations.
+When performing a fit, some values of the model function might have already been determined in
+previous experiments.
+Those results and uncertainties can then be used to constrain the given parameters in a new fit.
+This eradicates the need to manually propagate the uncertainties on the final fit results, as
```
```suggestion
This eliminates the need to manually propagate the uncertainties on the final fit results, as
```

JohannesGaessler reviewed (approved):
This is basically good to merge now; the only problem were a few typos/grammatical mistakes half of which were actually introduced by my own suggestions.

cverstege commented:
Fixed all known issues, including line breaks for long formulas and replaced `\textbf` with `\bm` in formulas for "correct" boldness in math mode.
Going to rebase and merge after next force push.
